src/DurableTask.Netherite/StorageLayer/Faster/AzureBlobs/AzureStorageDevice.cs

```diff
@@ -516,31 +556,83 @@ async Task ReadFromBlobAsync(UnmanagedMemoryStream stream, BlobUtilsV12.PageBlob
                         true,
                         async (numAttempts) =>
                         {
-                            if (numAttempts > 0)
-                            {
-                                stream.Seek(offset, SeekOrigin.Begin); // must go back to original position before retrying
-                            }
+                            stream.Seek(offset, SeekOrigin.Begin);
 
-                            if (length > 0)
-                            {
-                                var client = (numAttempts > 1 || length == MAX_DOWNLOAD_SIZE) ? blob.Default : blob.Aggressive;
+                            long requestedLength = length;
 
-                                var response = await client.DownloadStreamingAsync(
-                                    range: new Azure.HttpRange(sourceAddress + offset, length),
-                                    conditions: null,
-                                    rangeGetContentHash: false,
-                                    cancellationToken: this.PartitionErrorHandler.Token)
-                                    .ConfigureAwait(false);
+                            if (readCap.HasValue && sourceAddress + offset + requestedLength > readCap.Value)
+                            {
+                                requestedLength = readCap.Value - (sourceAddress + offset);
 
-                                using (var streamingResult = response.Value)
+                                if (requestedLength <= 0)
                                 {
-                                    await streamingResult.Content.CopyToAsync(stream).ConfigureAwait(false);
+                                    requestedLength = 0;
                                 }
                             }
 
-                            if (stream.Position != offset + length)
+                            if (requestedLength > 0)
                             {
-                                throw new InvalidDataException($"wrong amount of data received from page blob, expected={length}, actual={stream.Position}");
+                                var client = (numAttempts > 1 || requestedLength == MAX_DOWNLOAD_SIZE) ? blob.Default : blob.Aggressive;
```

Comment:
can you comment on what this controls? What is `blob.Aggressive`?

Reply:
they control the length of the timeout (with newer storage SDK the timeout can no longer be passed as a parameter to the operation, since it is a constructor parameter for the client, so we need multiple clients to control the timeout length).

BTW this was not modified in this PR, the diff just does not show this well.